### ISSUE TITLE
A2C does not normalize either advantage estimate or returns ? -> Loss explosion

### ISSUE THREAD

**Issue by zuoxingdong** [openai/baselines#362]
I am benchmarking OpenAI baselines implementation of A2C in Acrobot-v1. There is a phenomenon 
 observed, because baselines original implementation does not normalize either advantage estimate before backprop the action head or normalize returns before backprop value head.

When testing with Acrobot-v1, it leads to loss explosion, (loss value becomes very large like 20000), then the policy cannot learn anything.

After adding normalization to the returns for the value head, the problem solved. 

So it seems the point is that when we allow large max time step for an episode, without normalization, the MSE loss with become very big ?

**Comment by opherlieber** [comment]
What policy did you use for the A2C, all the existing ones are uint8 pixel-based? I had a similar issue of value-loss explosion with CartPole using a custom MLP policy, and the issue was that the A2C assumes uint8 observations, so any floating-point observations for example between 0-1 are truncated and lost, possibly it's the same with Acrobot observation space?

Try changing the dtype in this line:
  mb_obs = np.asarray(mb_obs, dtype=np.uint8).swapaxes(1, 0).reshape(self.batch_ob_shape)
To:
  mb_obs = np.asarray(mb_obs, dtype=self.obs.dtype).swapaxes(1, 0).reshape(self.batch_ob_shape)

**Comment by doviettung96** [comment]
Dear guys,
I just got the same problem you described. I tried to implement A2C and run in RoboschoolFlagrunHarder-v1 (open ai roboshool) and just after 100k timesteps all of my loss (pg_loss, value_loss) blow up to 7e5 for example. Then, I used A2C from baselines and after 4M timesteps, an exception of Inf/ NaN global norm occured. I have used the same hyperparameters set up such as 16 workers with 5 timesteps each. The environment's max timestep is 1000.
I will try to normalize the advantage and see what happen.
Thanks.
